### PR TITLE
webgpu: Refactor binary ops

### PR DESCRIPTION
**Pull request by gyagp**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6498)

### CODE REVIEW

**Comment by gyagp** [comment]
@qjia7 @xhcao PTAL

**Review by qjia7 — APPROVED** [review]
LGTM after fixing the bots failures. Thanks.

**Review by xhcao — APPROVED** [review]
Please fix bots failed issues.

**Comment by gyagp** [comment]
@qjia7 @xhcao I refactored the binary op to merge all the implementations into one program. Please review the change.

**Review by Neilblaze — APPROVED** [review]

File: tfjs-backend-webgpu/src/binary_op_webgpu.ts [hunk 1]
```diff
@@ -15,48 +15,132 @@
...
-          let b = getBByOutputIndex(index);
-          setOutputAtIndex(index, binaryOperation(a, b));
-        }
+      this.dispatch = computeDispatch(
+          this.dispatchLayout, this.outputShape, this.workGroupSize,
+          [this.workPerThread, 1, 1]);
+      this.shaderKey =
+          `binaryShared_${op}_${lastDimensionSize}_${useSharedMemoryWithB}`;
+
+      const sharedIndexSnippet = lastDimensionSize > 1 ?
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/binary_op_webgpu.ts:69]: The `getUserCode()`'s purpose is to hide the complexity of generating the shader body and only be called once if the `shaderKey` doesn't change. However, in this change, you break this rule and execute code L69~L140 every time when `new BinaryOpProgram`. You probably need to move them into `getUserCode` to  keep similar with before.

Comment (gyagp) [reply tfjs-backend-webgpu/src/binary_op_webgpu.ts]: I see. Please take another look at the fix.

**Review by qjia7 — COMMENTED** [review]
Please 1) split the change of `tfjs-backend-webgpu/yarn.lock` to a separate PR if really needed.
            2) resolve the bots failures.
LGTM after above two comments are resolved.

**Pull request merged**